### PR TITLE
fix: --embedded-database does not work when using fuel-indexer binary directly

### PR DESCRIPTION
**Pull request by lostman**
### Description

This PR makes `fuel-indexer start` run an embedded `postgres` database when `--embedded-database` flag is specified.

When `fuel-indexer` exits, the embedded database is shut down. For this purpose, `fuel-indexer` now listens to `SIGHUP/INT/TERM` signals.

### Testing steps

`fuel-indexer`

1. `--embedded-database`
```
cargo run -p fuel-indexer -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --auth-enabled --auth-strategy jwt --jwt-secret $FUEL_JWT_TOKEN --jwt-issuer FuelLabs --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

2. `--embedded-database` and `--local-fuel-node`
```
cargo run -p fuel-indexer --features api-server,fuel-core-lib -- run --run-migrations --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database --local-fuel-node
```
Note: `fuel-core-lib` must be enabled.

3 `--embedded-database` and `--all-features`, `--no-default-features`

```
cargo run -p fuel-indexer --all-features -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

```
cargo run -p fuel-indexer --no-default-features -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

4. `fuel-api-server` `--all-features`, `--no-default-features`

```
cargo run -p fuel-indexer-api-server --all-features -- run --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```

```
cargo run -p fuel-indexer-api-server --no-default-features -- run --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```

And check it works by visiting:
http://localhost:29987/api/playground/fuel_examples/explorer_indexer

### Changelog

* start an embedded database when `--embedded-database` is specified
* listen for termination signals
* stop the embedded database when a termination signal is received
* cleanup: `forc index start` passes `--embedded-database` flag to `fuel-indexer` instead of calling `forc_postgres::commands::create::exec`
* change how handles are created for various features—no nested cfg's
* vendored `openssl`
    it's a fix for
    ```
    ./fuel-node: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
    ```
    which was suggested here:
    https://github.com/launchbadge/sqlx/issues/473
    "the issue is because procedural macro executed in a slightly different environment"
    I think we're running into the same situation with our Docker build—we build with `cargo chef` image, but run on `ubuntu:22.04`

### CODE REVIEW

File: packages/fuel-indexer/src/commands/run.rs [hunk 1]
```diff
@@ -57,56 +100,82 @@ pub async fn exec(args: IndexerArgs) -> anyhow::Result<()> {
 
     let service_handle = tokio::spawn(service.run());
 
+    // for graceful shutdown
```
Comment (ra0x3) [inline packages/fuel-indexer/src/commands/run.rs:103]: @lostman 

- Just a heads up here.
- I see the logic of how the service starts and is feature flagged has been changed
- Totally fine with that overall, but just an FYI that the manual testing to verify this works is gonna be a bit vigorous
- You will need to test:
  - The service starts with all/any features `--all-features`, `--no-default-features`
  - The service starts with `--local-fuel-node`
  - Native execution starts with all/any features `--all-features`, `--no-default-features`
  - Naive execution starts with `--local-fuel-node`
  - Docker builds still work
    - Hello-indexer example
    - Block-explorer example
    - Hello-indexer-native example
- And we'd need to update the PR testing steps to reflect this 
  - You can see something somewhat similar over in https://github.com/FuelLabs/fuel-indexer/pull/772

File: packages/fuel-indexer/src/commands/run.rs [hunk 2]
```diff
@@ -57,56 +100,82 @@ pub async fn exec(args: IndexerArgs) -> anyhow::Result<()> {
...
+
+    use tokio::signal::unix::{signal, Signal, SignalKind};
 
-        let _ = service_handle.await?;
+    let mut sighup: Signal = signal(SignalKind::hangup())?;
+    let mut sigterm: Signal = signal(SignalKind::terminate())?;
+    let mut sigint: Signal = signal(SignalKind::interrupt())?;
 
-        Ok(())
+    tokio::select! {
```
Comment (ra0x3) [inline packages/fuel-indexer/src/commands/run.rs:148]: 👌🏽

**Review by ra0x3 — CHANGES_REQUESTED** [review]
@lostman 

- Looks mostly OK
- Left some minor comments
- Left a bigger comment about the feature flagging / service start change is gonna require manual testing steps
  - Hopefully https://github.com/FuelLabs/fuel-indexer/issues will alleviate this a bit

File: plugins/forc-index/src/ops/forc_index_start.rs [hunk 3]
```diff
@@ -107,6 +76,15 @@ pub async fn init(command: StartCommand) -> anyhow::Result<()> {
 
         match database.as_ref() {
             "postgres" => {
+                if embedded_database {
+                    if postgres_host.is_some() {
+                        panic!(
+                            "Cannot specify postgres host when using embedded database."
```
[the remark this answers is not in the document]
Comment (ra0x3) [reply plugins/forc-index/src/ops/forc_index_start.rs:82]: @lostman  Post our conversation, you're correct :)

**Review by ra0x3 — COMMENTED** [review]
Spent some time reviewing this:

The following things work:
- Ctrl+C stops `cargo run --embedded-database`
- `forc index kill` does not stop forc-postgres (but still stops forc-index)
- Ctrl+C stops `./target/release/fuel-indexer`
- `./target/release/fuel-indexer --embedded-database` works
- `./target/release/fuel-indexer --no-default-features` starts & stops
- hello-native works 

I was not able to get the following things to work:
- explorer-indexer works via docker
- hello-world works via docker

When trying to run the dockerized examples I'm consistently getting the following error

```text
> docker compose up 

block-explorer-fuel-indexer-1  | ./fuel-indexer: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
block-explorer-fuel-indexer-1 exited with code 127
```

My local image was built with the following command:

```bash
docker build -t fuel-indexer/local:latest -f deployment/Dockerfile .
```

@lostman I'm assuming the Dockerized examples worked for you?

**Comment by ra0x3** [comment]
> @ra0x3 `forc index kill` uses `SIGKILL`, so it leaves the `postgres` process hanging.
> 
> We could switch to `SIGTERM` as a default and add a `-9` flag for those cases when the users need it.
> 
> I'll re-check the Docker builds.

@lostman Yep, we can definitely handle the postgres killing separately :) Lmk how the Docker thing works for you. @deekerno  Could you also test the docker parts of this when you get a minute? I've already verified the other parts are working 👌🏽

**Review by deekerno — COMMENTED** [review]
Docker examples run correctly without `--embedded-database`, but not with the flag. @lostman, did you happen to run the Docker examples with the `postgres` container commented out? I may be missing something here.

**Review by deekerno — DISMISSED** [review]

**Comment by deekerno** [comment]
> > Docker examples run correctly without `--embedded-database`, but not with the flag. @lostman, did you happen to run the Docker examples with the `postgres` container commented out? I may be missing something here.
> 
> What is the error you're encountering?

After speaking about it with @lostman, I totally forgot that the fix itself would not be a part of the Docker compose workflow because it hasn't been included as part of an image yet. 😅 so I'm fine with this. Re-reviewing now.

**Review by deekerno — APPROVED** [review]

**Review by ra0x3 — APPROVED** [review]
`hello-world` is working via docker now ✅

**Pull request merged**